src/utilities/nep_utilities.cuh

```diff
@@ -58,6 +59,27 @@ static __device__ void apply_ann_one_layer(
   energy -= b1[0];
 }
 
+static __device__ void apply_gnn_one_layer(
+  const int dim, int num_neighbors, const float* theta, float* q_i, float* q_j, float* q_out)
+{
+  // TODO also add weights f_c(r_ij)
+  int F = dim; // dimension of q_out, for now dim_out = dim_in
+  for (int nu = 0; nu < dim; nu++) {
+    // Atom i
+    for (int gamma = 0; gamma < dim; gamma++) {
+      q_out[nu] += q_i[gamma] * theta[gamma + F * nu];
+    }
+    // neighbor atoms j
+    for (int j = 0; j < num_neighbors; j++) {
+      for (int gamma = 0; gamma < dim; gamma++) {
+        q_out[nu] = q_j[gamma] * theta[gamma + F * nu];
```

Comment:
Here I think should be (note the `+=` instead of `=`)
```
q_out[nu] += q_j[j + gamma * neighbor_number] * theta[gamma + F * nu];
```
You may also pay attention to the indexing of `theta` (might be `gamma  + dim *nu` if `F` is not equal to `dim`.

Reply:
Hmm yeah, I guess it would be most efficient to switch it around then? So it becomes 

'
for (int nu = 0; nu < dim; nu++) {
    ...
    for (int gamma = 0; gamma < F; gamma++) {
            q_out[nu] += q_j[gamma] * theta[gamma + dim * nu];
    }
}
'
So that a single GPU thread accesses theta contigously. I'm guessing "regular" CPU-like contigous access would be the fastest when using a single GPU thread?
where  `nu` runs over all dimensions

Reply:
It is not about performance, but about correctness. If F > dim, `gamma + F * nu` will exceeds the size `F * dim` because `nu` runs from 0 to F -1.

Reply:
Ah right, so if theta is stored as a Fxdim matrix, then the loops should become:

```python
for (int nu = 0; nu < F; nu++) {
    // Atom i
    for (int gamma = 0; gamma < dim; gamma++) {
      q_out[nu] += q_i[gamma] * theta[gamma + dim * nu];
    }
    // neighbor atoms j
    for (int j = 0; j < num_neighbors; j++) {
      for (int gamma = 0; gamma < dim; gamma++) {
        q_out[nu] += q_j[j + gamma * num_neighbors] * theta[gamma + dim * nu];
      }
    }
```
right?